### PR TITLE
feat: add instruction extraction with sync and async example notebooks

### PR DESCRIPTION
**Pull request by jojortz**
**Feature**
- Add instruction extraction to any-parser sdk.
- Refactor some repeated functionality into helper functions in any_parser.py
- Add `pdf_to_key_value.ipynb` and `async_pdf_to_key_value.ipynb`
- Add `pre-commit` packages to `pyproject.toml` and instructions in README.md

**Testing**
- Ran automated unit tests, including `test_sync_extract_key_value` and `test_async_extract_key_value_and_fetch`
- Ensured `pdf_to_key_value.ipynb` and `async_pdf_to_key_value.ipynb` ran without errors

### CODE REVIEW

File: any_parser/any_parser.py [hunk 1]
```diff
@@ -172,6 +227,8 @@ def async_extract(
             process_type = ProcessType.FILE_REFINED_QUICK
         elif model == ModelType.ULTRA:
```
Comment (goldmermaid) [inline any_parser/any_parser.py:228]: can you remove ultra?

**Comment by goldmermaid** [comment]
Can you add extract_json and the notebooks into our docs: https://docs.cambioml.com/api-reference/example?

show users what the differences between `sync` and `async` and how to properly use them;

File: any_parser/any_parser.py (hunk 1 above)
Comment (CambioML) [reply any_parser/any_parser.py]: Removed with latest commit.

Comment (jojortz) [reply any_parser/any_parser.py]: Removed in latest commit

File: any_parser/any_parser.py [hunk 2]
```diff
@@ -137,6 +136,62 @@ def extract(
         else:
             return f"Error: {response.status_code} {response.text}", None
 
+    def extract_json(
+        self,
+        file_path: str,
+        extract_instruction: Dict,
+    ) -> Tuple[str, str]:
+        """Extract json in real-time.
```
Comment (CambioML) [inline any_parser/any_parser.py:144]: This looks like a GenAI low quality dosstring, you should not just trust GenAI auto completely output, but given model models about extract what json based on what?

File: any_parser/any_parser.py [hunk 3]
```diff
@@ -190,33 +244,58 @@ def async_extract(
...
-            return f"Error: {response.status_code} {response.text}"
+        # If response successful, upload the file
+        return self._upload_file_to_presigned_url(file_path, response)
+
+    def async_extract_json(
+        self,
+        file_path: str,
+        extract_instruction: Dict,
+    ) -> str:
+        """Extract data asynchronously.
```
Comment (CambioML) [inline any_parser/any_parser.py:255]: Same. This what does extract data means in the docstring.

**Comment by CambioML** [comment]
> WAIT TO MERGE UNTIL THIS PR IS MERGED: [CambioML/any-parser-realtime-cdk#76](https://github.com/CambioML/any-parser-realtime-cdk/pull/76)
> 
> **Feature**
> 
>     * Add instruction extraction to any-parser sdk.
> 
>     * Refactor some repeated functionality into helper functions in any_parser.py
> 
>     * Add `pdf_to_json.ipynb` and `async_pdf_to_json.ipynb`
> 
> 
> **Testing**
> 
>     * Testing using notebooks, tested more extensively on backend

what does `tested more extensively on backend` means? Let's be explicit regarding what is tested to help us gain confidence.

File: any_parser/any_parser.py [hunk 4]
```diff
@@ -137,6 +136,62 @@ def extract(
         else:
             return f"Error: {response.status_code} {response.text}", None
 
+    def extract_json(
```
Comment (CambioML) [inline any_parser/any_parser.py:139]: This looks to me a very bad name. Especially, extract required no prompt while extract_json requires prompt. Then, suddently, you start to return json. Logically, I do not know why an extract_instruction starts to get this extract to extract_json. You should re-consider on naming this.

File: any_parser/any_parser.py [hunk 5]
```diff
@@ -190,33 +244,58 @@ def async_extract(
...
-                        return f"Error: {upload_resp.status_code} {upload_resp.text}"
-                return file_id
-            except json.JSONDecodeError:
-                return "Error: Invalid JSON response"
-        else:
-            return f"Error: {response.status_code} {response.text}"
+        # If response successful, upload the file
+        return self._upload_file_to_presigned_url(file_path, response)
+
+    def async_extract_json(
```
Comment (CambioML) [inline any_parser/any_parser.py:250]: same here.

File: any_parser/any_parser.py [hunk 6]
```diff
@@ -258,15 +340,53 @@ def async_fetch(
...
+            return f"Error: File does not exist: {file_path}"
+
+        if file_extension not in SUPPORTED_FILE_EXTENSIONS:
+            supported_types = ", ".join(SUPPORTED_FILE_EXTENSIONS)
+            return f"Error: Unsupported file type: {file_extension}. Supported file types include {supported_types}."
+
+    def _check_result_type(self, result_type: str) -> None:
+        if result_type not in RESULT_TYPES:
+            valid_result_types = ", ".join(RESULT_TYPES)
+            return f"Invalid result type: {result_type}. Supported `result_type` types include {valid_result_types}."
```
Comment (CambioML) [inline any_parser/any_parser.py:392]: Refactor to a utils.py to improve readability and no need to use _ then.

**Comment by jojortz** [comment]
> > WAIT TO MERGE UNTIL THIS PR IS MERGED: [CambioML/any-parser-realtime-cdk#76](https://github.com/CambioML/any-parser-realtime-cdk/pull/76)
> > **Feature**
> > ```
> > * Add instruction extraction to any-parser sdk.
> > 
> > * Refactor some repeated functionality into helper functions in any_parser.py
> > 
> > * Add `pdf_to_json.ipynb` and `async_pdf_to_json.ipynb`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **Testing**
> > ```
> > * Testing using notebooks, tested more extensively on backend
> > ```
> 
> what does `tested more extensively on backend` means? Let's be explicit regarding what is tested to help us gain confidence.

Updated comment with details on latest testing, which added unit tests to test.py for this key-value extract

File: any_parser/any_parser.py (hunk 2 above)
Comment (jojortz) [reply any_parser/any_parser.py]: Updated with latest commit

File: any_parser/any_parser.py (hunk 3 above)
Comment (jojortz) [reply any_parser/any_parser.py]: Updated with latest commit

File: any_parser/any_parser.py (hunk 4 above)
Comment (jojortz) [reply any_parser/any_parser.py]: updated to `extract_key_value`

File: any_parser/any_parser.py (hunk 5 above)
Comment (jojortz) [reply any_parser/any_parser.py]: updated to `extract_key_value`

File: any_parser/any_parser.py (hunk 6 above)
Comment (jojortz) [reply any_parser/any_parser.py]: moved to utils.py in latest commit

**Review by CambioML — APPROVED** [review]
LGTM -R

**Pull request merged**